### PR TITLE
isolated PR for seq_par with autograd fn to have more control over op…

### PR DESCRIPTION
**Pull request by ngoyal2707**
custom autograd function from transformer block to have more control over which ops to recompute.
I know the code is ugly and adds another codepath for model :( 

I think maybe we can just kill model parallel modules at some point?

### CODE REVIEW

File: metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates. All Rights Reserved.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+import torch
+
+import fused_layer_norm_cuda
```
Comment (suchenzang) [inline metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py:9]: Is this coming from apex installation?  Test are failing to import this...

File: metaseq/model_parallel/models/transformer_lm.py [hunk 2]
```diff
@@ -52,6 +52,18 @@ def build_model(cls, args, task):
...
+        if getattr(args, "sequence_parallel", False):
+            assert (
+                getattr(args, "model_parallel_size", 1) > 1
+            ), "--sequence-parallel only works when --model-parallel-size is greater than 1"
+            assert (
+                getattr(args, "dropout", 0.0) == 0.0
+            ), "havent yet tested if rng states are correct for dropout with seq_parallel"
+            assert (
+                getattr(args, "activation_fn", 'gelu') == 'gelu'
+            ), "For now only supports gelu"
```
Comment (suchenzang) [inline metaseq/model_parallel/models/transformer_lm.py:64]: Will there be a follow-up PR to add for swiglu? 😅

File: metaseq/modules/transformer_decoder_layer.py [hunk 3]
```diff
@@ -199,6 +200,18 @@ def forward(
         Returns:
             encoded output of shape `(seq_len, batch, embed_dim)`
         """
+        if getattr(self.args, "sequence_parallel", False):
```
Comment (suchenzang) [inline metaseq/modules/transformer_decoder_layer.py:203]: To confirm I'm following...
* We are still calling into ModelParallelTransformerDecoderLayer's build_fc1, build_fc2, etc. methods.
* These components (fc1, fc2, etc.) are then passed into `SequeuceParallelTransformerBlock`
* Killing model parallel modules going forward would imply that all models will be trained with sequence parallel

Separately, have a nit about naming since we separately have `TransformerDecoderMultiLayerBlockModule` floating around - is that still needed? If not, can open a task to remove (assume fsdp wrap every layer?) and keep the current `SequeuceParallelTransformerBlock` naming.

File: metaseq/models/transformer_decoder.py [hunk 4]
```diff
@@ -422,6 +422,7 @@ def extract_features(
                 incremental_state=incremental_state,
                 self_attn_mask=self_attn_mask,
                 self_attn_padding_mask=self_attn_padding_mask,
+                recompute_fc1=(idx < getattr(self.args, "recompute_fc1_num_layers", 0)),
```
Comment (suchenzang) [inline metaseq/models/transformer_decoder.py:425]: Is this.. the first N layers by design? Sorry, don't have much context here.

File: metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+        )
+        grad_k = grad_k.transpose(0, 1).contiguous().view(seq_len, bsz, -1)
+        grad_kvq_proj_output = torch.cat([grad_k, grad_v, grad_q], dim=-1)
+        return grad_kvq_proj_output
+
+    @staticmethod
+    def forward(
+        ctx,
+        input,
+        kvq_proj_weight,
```
Comment (suchenzang) [inline metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py:87]: I think I might've missed a transpose that happened somewhere..? `self.self_attn.qkv_proj.weight` is passed in here and it seems like this is intentionally kvq for a reason...

File: metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+        q  = q.view(seq_len, -1, head_dim)
+        k  = k.view(seq_len, -1, head_dim)
+        v  = v.view(seq_len, -1, head_dim).transpose(0, 1)
+
+        attn, _ = SequeuceParallelTransformerBlock.forward_mha(q, k, v, bsz, seq_len, head_dim, embed_dim_per_partition, dtype)
+
+        out_proj_out = torch.matmul(attn, out_proj_weight.t())
+        out_proj_out = _reduce_scatter_along_first_dim(out_proj_out)
+
+        # out_proj_out = out_proj_out + residual
```
Comment (suchenzang) [inline metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py:127]: Nit: remove?

File: metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+        out_proj_out = out_proj_out + residual
+
+        # Take out residual connection for FFN
```
Comment (suchenzang) [inline metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py:130]: Nit: `out_proj_out` seems to contain residual at this point, which gets passed into FFN... how residual get "taken out" here?

File: metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py (hunk 5 above)
Comment (suchenzang) [reply metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py:87]: Ah this looks like just a naming thing? Previously we had `kvq, _ = self.qkv_proj(query)`, which is just Line 113 below

File: metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py (hunk 7 above)
Comment (suchenzang) [reply metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py:130]: Oh maybe `out_proj_out = out_proj_out + residual` is supposed to be removed in Line 128?

Comment (suchenzang) [reply metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py:130]: Oh nvm, taken out == saved

File: metaseq/modules/transformer_decoder_layer.py (hunk 3 above)
Comment (ngoyal2707) [reply metaseq/modules/transformer_decoder_layer.py:208]: no we need to kill "TransformerDecoderMultiLayerBlockModule", I can send a separate PR for that

File: metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py (hunk 5 above)
Comment (ngoyal2707) [reply metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py:87]: makes sense, moved the comment up

File: metaseq/models/transformer_decoder.py (hunk 4 above)
Comment (ngoyal2707) [reply metaseq/models/transformer_decoder.py:425]: will add comment about it

**Review by suchenzang — APPROVED** [review]
Approving to unblock, assuming some of the tests eventually pass and... hoping that all the linear algebra matrix derivative transposes are just right 😅

File: metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py (hunk 1 above)
Comment (stephenroller) [reply metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py:9]: hm, why isn't it megatron.fused_kernels.fused_layer_norm?

Comment (ngoyal2707) [reply metaseq/model_parallel/modules/sequence_parallel_transformer_layer.py:9]: sorry its coming from apex, https://github.com/NVIDIA/apex/blob/master/apex/normalization/fused_layer_norm.py#L36-L37
